x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-DeltaAggregator.java.st

```diff
@@ -0,0 +1,250 @@
+            ) {
+                for (int i = 0; i < positionCount; i++) {
+                    final var groupId = selected.getInt(i);
+                    final var state = groupId < states.size() ? states.get(groupId) : null;
+                    if (state != null) {
+                        samples.beginPositionEntry();
+                        samples.appendLong(state.valuesSeen);
+                        samples.endPositionEntry();
+                        timestamps.beginPositionEntry();
+                        timestamps.appendLong(state.lastTimestamp);
+                        if (state.valuesSeen > 1) {
+                            timestamps.appendLong(state.firstTimestamp);
+                        }
+                        timestamps.endPositionEntry();
+
+                        values.beginPositionEntry();
+                        values.append$Type$(state.lastValue);
+                        if (state.valuesSeen > 1) {
+                            values.append$Type$(state.firstValue);
+                        }
+                        values.endPositionEntry();
+                    } else {
+                        samples.appendLong(0L);
+                        timestamps.appendNull();
+                        values.appendNull();
+                    }
+                }
+                blocks[offset] = samples.build();
+                blocks[offset + 1] = timestamps.build();
+                blocks[offset + 2] = values.build();
+            }
+        }
+
+        Block evaluateFinal(IntVector selected, GroupingAggregatorEvaluationContext evalContext) {
+            int positionCount = selected.getPositionCount();
+            try (DoubleBlock.Builder rates = evalContext.blockFactory().newDoubleBlockBuilder(positionCount)) {
+                for (int p = 0; p < positionCount; p++) {
+                    final var groupId = selected.getInt(p);
+                    final var state = groupId < states.size() ? states.get(groupId) : null;
+                    if (state == null || state.valuesSeen < 2) {
+                        rates.appendNull();
+                        continue;
+                    }
+                    if (evalContext instanceof TimeSeriesGroupingAggregatorEvaluationContext tsContext) {
+                        // At this point we want to apply extrapolation
+                        var rangeStart = tsContext.rangeStartInMillis(groupId);
+                        var rangeEnd = tsContext.rangeEndInMillis(groupId);
+                        if (state.lastTimestamp - state.firstTimestamp == 0) {
+                            rates.appendNull();
+                            continue;
+                        }
+                        double startGap = state.firstTimestamp - rangeStart;
+                        final double averageSampleInterval = (state.lastTimestamp - state.firstTimestamp) / state.valuesSeen;
```

Comment:
I think we should cast `(state.lastTimestamp - state.firstTimestamp)` to a double before the division